RFC: API Change: usb: Make users call usb_enable. Provide global status callback.

eobalski opened the pull request:
This PR is for resolving point 1 from #13403
Closes #21419

Commit:
usb: samples: Application calling usb_enable by itself

User app is reponsible for issuing usb_enable and
making USB hardware operative.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

Commit:
usb: api: Add user device status callback

By this commit user gets possibility to register USB
device satutus callback. This callback represents device state
and is added so user could know what happend to USB device.

Callback is registered by providing it to usb_enable()
USB api is extended by this callback handler.

Samples using using USB are by default provide no callback
and the usb_enable() is called with NULL parameter.

Status callback registered by hid class is deleted as now
USB device has global callback for all classes within device.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

carlescufi reviewed (commented):
I think this is a reasonable way of solving this.
We should in theory deprecate `usb_enable()` since that is a public API, but in practice this is probably never been used before.

@jfischer-phytec-iot and @finikorg can you also state your opinion?

pabigot commented on include/usb/usb_device.h:
```diff
@@ -96,6 +96,31 @@ struct usb_setup_packet {
+	USB_DEVICE_CONFIGURED,
+	/** USB device has been suspended due to bus inactivity */
+	USB_DEVICE_SUSPENDED,
+	/** USB device number of states */
```
Please remove `USB_DEVICE_NUM`.  Unless there's a reason why we need to iterate through enum values there's no need to name the upper bound, and doing so is confusing as there is no `NUM` USB device state.

pabigot commented on subsys/usb/usb_device.c:
```diff
@@ -150,6 +150,8 @@ static struct usb_dev_priv {
 	usb_request_handler custom_req_handler;
 	/** USB stack status clalback */
 	usb_dc_status_callback status_callback;
+	/** USB user status clalback */
```
s/clalback/callback/g

pabigot commented on subsys/usb/usb_device.c:
```diff
@@ -971,6 +986,10 @@ static void forward_status_cb(enum usb_dc_status_code status, const u8_t *param)
 		}
 	}
 
+	if (usb_dev.user_status_callback) {
```
Was this intended to map `enum usb_dc_status_code` to `enum usb_state`?  If not, we can drop the commit that adds that enum.

jfischer-phytec-iot replied:
[usb_dc_status_code](https://github.com/zephyrproject-rtos/zephyr/blob/master/include/drivers/usb/usb_dc.h#L41) should be reordered

emob-nordic replied:
Adding this enum I intended to provide to user the **state** enum type which will match states described in spec and nothing more. The callback is still using **statuses** as my intention was that the user is responsible for controlling the state.

emob-nordic replied:
> [usb_dc_status_code](https://github.com/zephyrproject-rtos/zephyr/blob/master/include/drivers/usb/usb_dc.h#L41) should be reordered

Do You mean the order of statuses should match those from [usb_state](https://github.com/zephyrproject-rtos/zephyr/pull/20375/files#diff-0a5e6a18914d2e2bf1bf9595e8ab483eR105) ?

jfischer-phytec-iot replied:
Yes, otherwise it will not work

emob-nordic replied:
Right now I dont think there is a need to change anything in the  [usb_dc_status_code](https://github.com/zephyrproject-rtos/zephyr/blob/master/include/drivers/usb/usb_dc.h#L41) as I agreed to remove usb_device_state (justification below)

jfischer-phytec-iot replied:
:confused: But was it not what @carlescufi and you wanted?

carlescufi replied:
What we wanted is to expose the Device States from the spec to the application. I disagree with @pabigot here, and I think we should expose them all. I see no need to just expose a subset.

pabigot replied:
To clarify: my objection was to adding `enum usb_state` as a distinct type but not providing the translation from the existing DC states to them when passing them to the callback.  If you want to do the translation, that's fine, but from what I understand it would only work on Nordic because the other implementations can't provide the necessary detail.

Also, I do believe that "disconnected" is a state I need to be informed about (distinct from suspended), even if the spec doesn't consider it a named "device state".

pabigot reviewed (changes requested):
As implemented the new USB device state enumeration is not used.  It should be removed, or the DC states mapped to device states.  This is the blocker, the rest is commentary.

I still don't like the fact that there's a single callback, which means every application that uses USB has to track the state and fan it out to modules that care about it.  I'd prefer to have something like:
```
struct usb_state_callback_node {
    sys_snode_t node;           /* internal use */
    usb_dc_status_callback cb;  /* initialize to user callback */
};

/** Link in a state-change callback.
 * @param node pointer to a persistent object initialized to contain a user callback.
 * Return: The current USB state, or a negative error code. */
int usb_state_add_callback(struct usb_state_callback_node *node);

/** Unlink a state-change callback. */
int usb_state_remove_callback(struct usb_state_callback_node *node);
```

Note that once there's an option to add a callback after USB has been initialized we need a way to ensure that the application tracks the state correctly: either by returning the state, or by having the callback invoked with an initial state when it's registered.

But I guess that's complex, and if nobody else thinks it's worthwhile I can live with doing the status fan-out in each application.


For the record the sequence I get on startup on NRF52840 is:
```
Status: 2 CONNECTED
Status: 5 SUSPEND
Status: 6 RESUME
Status: 1 RESET
Status: 1 RESET  // twice?
Status: 3 CONFIGURED
```

When the USB is unplugged I get:
```
Status: 5 SUSPEND
Status: 4 DISCONNECTED
```

If I plug in a power-only line I get:
```
Status: 2 CONNECTED
Status: 5 SUSPEND
```

which is a little odd.  It's not obvious how those map to device states (which should be extended to include a "DISCONNECTED" state, if that enum is going to be used).

jfischer-phytec-iot commented:
@carlescufi @emob-nordic @finikorg After reviewing this and #20011 I think we should not introduce USB_DEVICE_AUTO_ENABLE and drop 7bc65fabccc57cb8d4235ac9a8d2a604b33ba856

jfischer-phytec-iot commented on include/usb/usb_device.h:
```diff
@@ -96,6 +96,31 @@ struct usb_setup_packet {
+ * State Codes gives an overview to user application what is USB device state
+ * For more information refer to chapter 9 from USB 2.0 spec.
+ */
+enum usb_state {
```
usb_device_state?

emob-nordic replied:
Yes, will update it.

jfischer-phytec-iot commented:
> Note that once there's an option to add a callback after USB has been initialized we need a way to ensure that the application tracks the state correctly: either by returning the state, or by having the callback invoked with an initial state when it's registered.

This callback is used for notification only. The stack does not have to care about what application does with that. btw, I think a message queue or mailbox would be a better solution, but needs more work.

emob-nordic commented on include/usb/usb_device.h:
```diff
@@ -96,6 +96,31 @@ struct usb_setup_packet {
+	USB_DEVICE_DEFAULT,
+	/** USB device has been assigned a unique address */
+	USB_DEVICE_ADDRESS,
+	/** USB device was chosen a configuration */
```
Device can be resumed from suspend and end up in not only configured state. From Suspend You can either enter Default, Address or Configured State. I had a trouble describing Configured state in one sentence because as I refered to Table 9-1 from usb 2.0 spec (chapter 9)
_"Device is attached to the USB, powered, has been reset, has a unique address, is configured, and is not suspended. The host may now use the function provided by the device."_

Maybe it would be better something like:
_"USB device is configured and the host may use the function provided by the device"_ ?

jfischer-phytec-iot replied:
No, usual it is Configured->Suspend->Configured. Not at all something like Configured->Suspend->Address.

emob-nordic replied:
I meant depending in which state the device was before after entering Suspended it can either end up in x = (Powered, Default, Address, Configured). It's possible to have those transitions:
x -> Suspended -> x

emob-nordic commented on subsys/usb/usb_device.c:
```diff
@@ -931,6 +933,19 @@ static void usb_register_status_callback(usb_dc_status_callback cb)
+ *
+ * @param [in] cb Callback function pointer
+ */
+static void usb_register_user_status_callback(usb_dc_status_callback cb)
```
I was following the existing style it is present to this file. Same was done to [usb_register_status_callback](https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/usb/usb_device.c#L929). I wanted to be consistent but of course it could be simplified, I do not insist.

jfischer-phytec-iot replied:
Please simplify :-)

emob-nordic commented:
>As implemented the new USB device state enumeration is not used. It should be removed, or the DC states mapped to device states. This is the blocker, the rest is commentary.

The usb_device_state was prepared for users so they will not create their own enum for this. Nothing more.

I understand that ideally user_state_callback should use usb_device_state as an input parameter but:
- from what I checked it would only apply for nordic (some vendors forward only 2 statuses which is weird but it is done like this and its not my intention here to correct drivers).

Ideally We would have field device_state in usb_dev singleton and then based on **status** callback we could change the **state** internally in the device and then call  proper user callback to keep him informed about the **state**.

Just to clarify the view of this PR. What basically happens here is:
- throw away the hid_ops.status_cb and makes it global per device, not per class as this should never be class specific.

> which is a little odd. It's not obvious how those map to device states (which should be extended to include a "DISCONNECTED" state, if that enum is going to be used).

usb_state will not include DISCONNECTED state as its not consistent with the spec and for this I would not agree with You.

pabigot commented:
> Just to clarify the view of this PR. What basically happens here is:
> 
> * throw away the hid_ops.status_cb and makes it global per device, not per class as this should never be class specific.

That's a good goal.

> The usb_device_state was prepared for users so they will not create their own enum for this. Nothing more.
> usb_state will not include DISCONNECTED state as its not consistent with the spec

I don't see the value in providing this enum if it's not going to be handled by the driver.  To this point AFAIK no classes have had any use for these states.  Even if they did, your point that some vendor implementations don't provide enough detail through DC states to distinguish them suggests they aren't generally useful.

Whatever receives the state changes will still have to interpret the DC states provided in the callback and translate them into whatever states are important.  For me that's "connected/working" and "disconnected/not working".  I do not care about the finer-grained states in `enum usb_state`.

I need a compelling reason why `enum usb_state` should remain, or fairly consistent agreement from the "real" USB maintainers that it should be here even if it's not used, before I withdraw the NAK.

Commit:
usb: Add mutex to usb_enable

The purpose of this commit is to prevent user from
calling usb_enable() twice from different contexes.

If for example user configures composite device with hid
and USB uart console (CONFIG_USB_UART_CONSOLE=y)
then introduced mutex will prevent from calling
usb_enable twice and thus lead to undefined behaviour of
usb driver controller.

usb_enable shall always be called once as it refers to usb driver
and for now only one USB driver instance is supported in Zephyr.
This mechanism ensures that.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

emob-nordic commented:
> I need a compelling reason why `enum usb_state` should remain, or fairly consistent agreement from the "real" USB maintainers that it should be here even if it's not used, before I withdraw the NAK.

I've discussed it internally and I agree in this matter. It's not common to all drivers (unfortunate) so shall not be there.

finikorg commented on samples/subsys/usb/hid-cdc/src/main.c:
```diff
@@ -620,6 +626,13 @@ void main(void)
 	usb_hid_register_device(hid1_dev, hid_kbd_report_desc,
 				sizeof(hid_kbd_report_desc), &ops);
+
+	ret = usb_enable(status_cb);
```
So to use this callback we shall call usb_enable(cb) always?

emob-nordic replied:
Yes. If callback is needed then USB_DEVICE_AUTO_ENABLE=n and call usb_enable(cb) in the app.

jfischer-phytec-iot commented on samples/subsys/usb/hid-cdc/src/main.c:
```diff
@@ -620,6 +626,13 @@ void main(void)
 				sizeof(hid_kbd_report_desc), &ops);
+
+	ret = usb_enable(status_cb);
+	if (ret != 0) {
```
`if (usb_enable(status_cb))`?

pabigot replied:
Personal opinion: yeah, it could be[*], but that would make it slightly more inconvenient to add diagnostics to display exactly why it failed when debugging or maintaining.

[*] well, actually this:

```
if (usb_enable(status_cb) != 0) {
```

since we're courting MISRA these days and treating `int` values as booleans is not acceptable.

jfischer-phytec-iot commented on samples/subsys/usb/hid-mouse/src/main.c:
```diff
@@ -261,12 +262,16 @@ void main(void)
+				NULL);
+
+	ret = usb_enable(status_cb);
+	if (ret != 0) {
```
`if (usb_enable(status_cb))`?

jfischer-phytec-iot commented on samples/subsys/usb/hid/src/main.c:
```diff
@@ -109,15 +109,22 @@ static void protocol_cb(u8_t protocol)
 	LOG_DBG("Starting application");
 
+	ret = usb_enable(status_cb);
+	if (ret != 0) {
```
`if (usb_enable(status_cb))`?

pfalcon commented on samples/net/wpanusb/src/wpanusb.c:
```diff
@@ -417,6 +418,11 @@ void main(void)
 
 	radio_api = (struct ieee802154_radio_api *)ieee802154_dev->driver_api;
 
+	ret = sb_enable(NULL);
```
What is sb_enable()? This may be an illustration why changes like this don't scale.

emob-nordic replied:
Sorry for that. I have to remove it by mistake when rebasing.

carlescufi commented on subsys/usb/usb_device.c:
```diff
@@ -971,6 +975,10 @@ static void forward_status_cb(enum usb_dc_status_code status, const u8_t *param)
 	}
 
+	if (usb_dev.user_status_callback) {
+		usb_dev.user_status_callback(status, param);
```
This should be done after the built-in classes have been informed

carlescufi reviewed (approved):
I see no other clean way to do this since we are already exposing the application to `usb_dc_status` with `usb_cfg_data`.

pabigot commented on subsys/usb/usb_device.c:
```diff
@@ -1542,18 +1544,24 @@ int usb_enable(void)
+	 * This should only be called once.
+	 */
+	LOG_DBG("lock usb_enable_lock mutex");
+	k_mutex_lock(&usb_enable_lock, K_FOREVER);
```
The above check for `usb_dev.enabled` must be moved after the mutex is acquired, or two threads could pass that state, then sequentially both initialize the device.

emob-nordic replied:
Good point. I missed this case.

jfischer-phytec-iot replied:
Guess it was my fault too.

pabigot reviewed (changes requested):
The `usb_dev.enabled` check must be moved into the mutex region.

emob-nordic commented:
@pabigot, @jfischer-phytec-iot  I've updated suggested changes. Please check it once again and approve if it's fine.

pabigot commented on subsys/usb/usb_device.c:
```diff
@@ -1538,22 +1540,28 @@ int usb_enable(void)
+	k_mutex_lock(&usb_enable_lock, K_FOREVER);
+
 	if (usb_dev.enabled == true) {
 		return 0;
```
No, this will produce deadlocks for anything that subsequently invokes `usb_enable()`.  The body has to be:
```
    ret = 0;
    goto out;
```
where the label `usb_enable_error` is renamed to `out`.

emob-nordic replied:
Done.

pabigot reviewed (dismissed):
Need to ensure mutex is unlocked when leaving the mutex region.

pabigot commented on subsys/usb/usb_device.c:
```diff
@@ -1600,17 +1608,20 @@ int usb_enable(void)
 	usb_dev.enabled = true;
 
-	return 0;
+usb_enable_error:
```
I would like this to be renamed to `out` as it is the exit path for success as well.

pabigot commented on subsys/usb/usb_device.c:
```diff
@@ -1600,17 +1608,20 @@ int usb_enable(void)
+		goto usb_enable_error;
 	}
 
 	usb_dev.enabled = true;
```
To be consistent with the documented behavior of `usb_enable()` this should be added, in case some of the intermediate functions returned a positive value:

    ret = 0;

emob-nordic replied:
To be honest I dont get it. if failure, then return success?

jfischer-phytec-iot replied:
API documented behavior for `usb_enable()` is  _"0 on success, negative errno code on fail."_, so it is undefined for (ret > 0) :-)

But as we are here, should we cleanup and change this` if (ret < 0) `to `if (ret)`?
- `static int usb_vbus_set(bool on);` returns 0 on success, negative errno code on fail
- `int usb_dc_attach(void); `returns 0 on success, negative errno code on fail
- `int usb_dc_ep_configure(const struct usb_dc_ep_cfg_data * const cfg);` returns 0 on success, negative errno code on fail
- `int usb_dc_ep_set_callback(const u8_t ep, const usb_dc_ep_callback cb);` returns 0 on success, negative errno code on fail
- `static int composite_setup_ep_cb(void); `returns 0 on success, negative errno code on fail
- `int usb_dc_ep_enable(const u8_t ep);` returns 0 on success, negative errno code on fail

pabigot replied:
> API documented behavior for `usb_enable()` is _"0 on success, negative errno code on fail."_, so it is undefined for (ret > 0) :-)

Exactly.  Failing to describe the complete value range of input and output parameters is pervasive problem in Zephyr.

> But as we are here, should we cleanup and change this`if (ret < 0)`to `if (ret)`?

Changing to `if (ret != 0)` would be acceptable as it satisfies the return guarantee--as long as none of those functions return positive.  Making a maintainer search for the documented (and implemented) behavior of all those subordinate functions to confirm the API  is not reasonable IMO, hence the change request.

jfischer-phytec-iot replied:
@emob-nordic Do you agree with "Changing to if (ret != 0) would be acceptable" ?

emob-nordic replied:
Yes, now I get the point. I will update it in a minute.

pabigot commented on subsys/usb/usb_device.c:
```diff
@@ -1645,8 +1645,6 @@ static int usb_device_init(struct device *dev)
 
 	usb_set_config(device_descriptor);
 
-	usb_enable();
```
Here's one where I'm not responsible for this, so I won't block, but: This is unilaterally changing existing behavior of a subsystem, requiring code to change so the application continues to work.  There's no documentation of this change.  At a minimum it should be part of a release note on changed API.  Somebody should figure out how to make that work.

emob-nordic replied:
The story behind this change starts from Nordic application team. There was a problem with missed status_callbacks in [hid_ops.status_cb](https://github.com/zephyrproject-rtos/zephyr/blob/master/include/usb/class/usb_hid.h#L85) Two issues here:
1) status callback should never been associated with class.
2) usb_enable() calls usb_dc_attach() and by that call interrupts are enabled. Because it happens on SYS_INIT and user app is registering HID callbacks by itself (after system is woken up), it is possible to get the interrupt from usbd and in particular miss some statuses in the APP.

By this change we let the user to control usb_enable function and by that avoid those misses.

pabigot replied:
By this change you **require** the user to control `usb_enable()`, causing all existing USB applications to break until they are modified to enable USB explicitly, just as you've done for the examples.  Or am I wrong about that?

I understand the reason for the change but a comment here in the PR doesn't help the maintainers of existing code figure out why their applications no longer work.

emob-nordic replied:
@pabigot I will update documentation in commit in which I am changing from _auto enable_ to _enable by user_ as well as PR message. Would this be sufficient?

pabigot replied:
The [documentation](https://docs.zephyrproject.org/latest/development_process/api_lifecycle.html#deprecated) has a process.

I was going to say something about it assuming the API can be deprecated (rather than changed immediately) and focuses on stable APIs, but I looked at the [API Status](https://github.com/zephyrproject-rtos/zephyr/projects/15) board and USB device is marked as stable.

@carlescufi does this change warrant an exception?

carlescufi replied:
@pabigot you raise a good point but here are the circumstances around this that I'd like to point out to warrant an exception in this particular case:

- The current behavior was inconsistent with other protocol stacks in Zephyr, and in particular with Bluetooth, that requires a `bt_enable()` call from the sample that controls the moment at which we power on the hardware.
- The `usb_enable()` function was almost useless in the form that existed until this PR, and we are now giving it a purpose.
- I agree we need to document this behavior in two places: The doxygen for `usb_enable()` itself and the release notes (2.2 release notes at this point).
- We will merge this change right in the beginning of the 2.2 release cycle, which give ample time to users that follow master to adapt.
- Deprecating the current behavior would not benefit users in any meaningful way, because of the current inconsistent and incorrect behavior of this `usb_enable()` API.
- For users of standard provided classes that expose an API, eg. HID there was a race condition with the current master behavior of auto-enabling the USB stack: the USB stack was enabled before the user had a chance to register a HID device, meaning that events could be lost in between.

pabigot reviewed (dismissed):
One blocking (ensuring `usb_enable()` returns zero as documented when it succeeds), one suggestion, and one note on the implicit, significant change in API that now requires the application to enable the device.

When I first started reviewing this PR I remember the intent was to have a Kconfig that controlled auto-enable.  I would be willing to approve this if that was restored and defaulted to auto-enable, or if there was a clear documented policy that allowed the change and explained how it would be communicated to users, but otherwise I'll stick to "won't reject".  I feel strongly that Zephyr must be much better about not breaking existing code even if the API is "unstable".

pabigot reviewed (commented):
I like (and need) the new capability, but can't approve the API breakage being merged with no attempt to follow the standard deprecation process or to extend the process to cover cases where deprecation is simply infeasible.

carlescufi commented:
As per the API meeting, @emob-nordic please bring back the Kconfig option to conditionally disable the auto-powerup of the USB stack into this PR.

jfischer-phytec-iot commented:
> As per the API meeting, @emob-nordic please bring back the Kconfig option in #20011 to conditionally disable the auto-powerup of the USB stack into this PR.

Disagree, that is not the problem, the cb argument breaks the API.

jfischer-no committed:
usb: initialize USB descriptor earlier

Initialize USB descriptor earlier.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-no reviewed (changes requested):
We forgot to rework netusb, please take a look/cherry-pick https://github.com/jfischer-phytec-iot/zephyr/tree/20375

emob-nordic commented:
> > As per the API meeting, @emob-nordic please bring back the Kconfig option in #20011 to conditionally disable the auto-powerup of the USB stack into this PR.
> 
> Disagree, that is not the problem, the cb argument breaks the API.
> But there is more to discuss, this PR breaks also netusb support.

I will leave this discussion for USB sync meeting today.

carlescufi commented:
@emob-nordic  please follow [the new procedure](https://docs.zephyrproject.org/latest/development_process/api_lifecycle.html#introducing-incompatible-changes) and let's not introduce the Kconfig option, taking the opportunity to clean this up since we are introducing incompatible changes anyway.

Commit:
doc: release notes: Update contents for 2.2.0

Update to API change in regards of USB changes.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

dbkinder commented on doc/releases/release-notes-2.2.rst:
```diff
@@ -27,6 +27,12 @@ Deprecated in this release
+
+  * Users are obligated to enable USB by themselves.
+    usb_enable() is taking one parameter which is callback function.
+    Callback funciton will be called whenever USB driver reports change.
```
``function`` misspelled

dbkinder reviewed (dismissed):
Seems like there should be some usage documentation changes for this (other than in the doxygen comments and release notes)...

emob-nordic commented:
> Seems like there should be some usage documentation changes for this (other than in the doxygen comments and release notes)...

What would You suggest? I was following the [process](https://docs.zephyrproject.org/latest/development_process/api_lifecycle.html#introducing-incompatible-changes) and hoped this is sufficient.

carlescufi commented on doc/releases/release-notes-2.2.rst:
```diff
@@ -27,6 +27,12 @@ Deprecated in this release
 
+* USB
+
+  * Users are obligated to enable USB by themselves.
```
I will give a shot at another description here @dbkinder and @emob-nordic:
------
* The `usb_enable()` function, which was previously invoked automatically by the USB stack, now needs to be explicitly called by the application in order to enable the USB subsystem.
* The `usb_enable()` function now takes a parameter,  `usb_dc_status_callback` which can be set by the application to a callback to receive status events from the USB stack. The parameter can also be set to `NULL` if no callback is required.

----

StefJar replied:
lovely - I like it. IMHO this is how it should be

carlescufi commented:
> Seems like there should be some usage documentation changes for this (other than in the doxygen comments and release notes)...

I proposed a rewritten section in the release notes. Beyond this and the doxygen changes, there is currently no documentation which explains to the users the USB subsystem usage, so I do not think any further changes are necessary until we add more documentation @dbkinder.
The change follows standard procedure as pointed out by @emob-nordic

joerchan commented on samples/bluetooth/hci_usb/src/main.c:
```diff
@@ -6,8 +6,16 @@
+	if (ret != 0) {
+		printk("Failed to enable USB");
+		return;
+	}
```
blank space after `}`
